src/graph/CanvasGraphLayouts.js

```diff
@@ -0,0 +1,326 @@
+                }
+            }
+        },
+        "ForceDirected2": {
+            decay: { alpha: 0.03, velocity: 0.6 },
+            xy: force_directed_forces.xy,
+            link: force_directed_forces.link,
+            manyBodyForce: force_directed_forces.manyBodyForce,
+            zoom_by_force: false,
+            init: function (data) {
+                //
+            },
+            post_init: function (data) {
+                //                for(var i = 0;i<500;i++){
+                //                    this.simulation.tick();
+                //                }
+                //                data.nodes.forEach(function (_node, _i, _arr) {
+                //                    _node.fx = _node.x;
+                //                    _node.fy = _node.y;
+                //                });
+            }
+        },
+    };
+    //    "ForceDirected2": {
+    //        decay:{alpha:0.0228,velocity:0.7},
+    //        xy:{
+    //            x:function() {return 0;},
+    //            y:function() {return 0;},
+    //            strength:function (d) {
+    //                return d.child_arr.length > 0 ? 1 : -0.1;
+    //            },
+    //            post:false
+    //        },
+    //        link:{
+    //            distance:100,
+    //            strength:function (d) {
+    //                if (d.target.is_primary) {
+    //                    return 200;
+    //                }
+    //                if (d.target.is_leaf) {
+    //                    return 130;
+    //                }
+    //                return 60;
+    //            }
+    //        },
+    //        manyBodyForce:{
+    //            strength:function (d) {
+    //                if(d.is_primary)return -400;
+    //                if (d.is_leaf) {
+    //                    return -200;
+    //                }
+    //                return -250;
+    //            }
+    //        },
+    //        zoom_by_force:false,
+    //        init: function (data) {
+    //            //
+    //        }
+    //    }
+
+    function createDagreGraph(data) {
+        var g = new dagre.graphlib.Graph({
+            multigraph: true,
+            compound: true,
+            directed: true
+        });
+        g.setGraph({
+            rankdir: this.hierarchyRankDirection(),
+            nodesep: this.hierarchyNodeSeparation(),
+            edgesep: this.hierarchyEdgeSeparation(),
+            ranksep: this.hierarchyRankSeparation()
+        });
+        g.setDefaultEdgeLabel(function () {
+            return {};
+        });
+        data.nodes.forEach(function (u) {
+            console.info("u.width:", u.width);
+            g.setNode(u.index, {
+                width: u.width,
+                height: u.height
+            });
+        });
+        data.links.forEach(function (u) {
+            g.setEdge(u.source.index, u.target.index);
+        });
+        return g;
+    }
+
+    return {
+        layouts: layouts
+    };
+}));
```

Comment:
Looks like duplicate code with the existing graph layouts?

Reply:
Yes, some duplicate code. V2 should be able to share layout code with SVG graph.